bazel/integration/test_runner/runner.ts

```diff
@@ -0,0 +1,215 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as fs from 'fs';
+import * as path from 'path';
+import * as tmp from 'tmp';
+import {addWritePermissionFlag, writeExecutableFile} from './file_system_utils';
+import {
+  PackageMappings,
+  readPackageJsonContents,
+  updateMappingsForPackageJson,
+} from './package_json';
+import {BazelFileInfo, resolveBazelFile, resolveBinaryWithRunfiles} from './bazel';
+import {debug} from './debug';
+import {
+  expandEnvironmentVariableSubstitutions,
+  getBinaryPassThroughScript,
+  prependToPathVariable,
+  runCommandInChildProcess,
+} from './process_utils';
+
+/**
+ * Test runner that takes a set of files within a Bazel package and copies the files
+ * to a temporary directory where it then executes a list of specified commands.
+ *
+ * Additionally, a list of NPM packages can be mapped to Bazel artifacts, allowing for
+ * locally-built first-party packages to be tested within such an integration test. The
+ * test runner will patch the top-level `package.json` of the test Bazel package for that.
+ */
+export class TestRunner {
+  constructor(
+    private readonly testFiles: BazelFileInfo[],
+    private readonly testPackage: string,
+    private readonly toolMappings: Record<string, BazelFileInfo>,
+    private readonly npmPackageMappings: Record<string, BazelFileInfo>,
+    private readonly commands: [[binary: string, ...args: string[]]],
+  ) {}
+
+  async run() {
+    const tmpDir = await this._getTmpDirectoryPath();
+    const toolMappings = await this._setupToolMappingsForTest(tmpDir);
+
+    await this._copyTestFilesToDirectory(tmpDir);
+    await this._patchPackageJsonIfNeeded(tmpDir);
+    await this._runTestCommands(tmpDir, toolMappings.binDir);
+  }
+
+  /**
+   * Gets the path to a temporary directory that can be used for running the integration
+   * test. The temporary directory will not be deleted as it is controlled by Bazel.
+   *
+   * In case this test does not run as part of `bazel test`, a system-temporary directory
+   * is being created, although not being cleaned up to allow for debugging.
+   */
+  private async _getTmpDirectoryPath(): Promise<string> {
+    // Bazel provides a temporary test directory itself when it executes a test. We prefer
+    // this when the integration test runs with `bazel test`. In other cases we want to
+    // provide a temporary directory that can be used for manually jumping into the
+    // directory. The Bazel test tmpdir is not guaranteed to remain so for debugging,
+    // when the test is run with `bazel run`, we use a directory we control.
+    if (process.env.TEST_TMPDIR) {
+      // Bazel itself cleans the temporary directory, so the direct cleanup
+      // function is a noop.
+      return process.env.TEST_TMPDIR;
+    }
+
+    return new Promise((resolve, reject) => {
+      tmp.dir(
+        {
+          template: 'ng-integration-test-XXXXXX',
```

Comment:
Where is this temp directory created?  In `/tmp`?

Reply:
yeah, that is the system temp directory, or whatever the `tmp` library maintains.